### ISSUE TITLE
New Repo: mink

### ISSUE THREAD

**Issue by mattmoor** [knative/community#335]
## Repo information

**Org:** knative-sandbox

**Repo:** mink

**Purpose (Description):**

tl;dr The thrust of this proposal is to move [`mink`](github.com/mattmoor/mink) into the Knative sandbox.

Mink started as a demonstration that you can leverage the Knative controller architecture to “shrink wrap” more opinionated installations of Knative (and downstream bits like Tekton!) with relatively small amounts of code.  Where our general bias upstream is for optionality, mink makes certain choices that limit install-time optionality, but without sacrificing the general Knative extensibility model.  For example, mink bundles [net-contour](https://github.com/knative-sandbox/net-contour) and [net-http01](https://github.com/knative-sandbox/net-http01), but could still be configured to use [net-istio](https://github.com/knative-sandbox/net-istio) and [net-certmanager](https://github.com/knative-sandbox/net-certmanager).

The key goals, which mink achieves, are:

1. To make installation as simple as a single `kubectl apply`, and
2. To run in as small a footprint as possible without sacrificing HA and scalability.

The intention of this proposal is **NOT** to pick a ”winning” opinion (for this reason I have had an aversion to mink living upstream since its inception!).  The goal is closer to the dynamic with KinD and minikube in Kubernetes (and to a lesser extent k3s under CNCF), to present *an* option, not *the* option.  The recent clarity around knative sandbox feels almost purpose-built for projects like this, and so I am proposing that we move mink under the Knative sandbox.

As the scope of mink has shifted over time, it is worth calling out the “bill of materials” that I see as a part of its initial scoping:

- Serving
  - net-contour
  - net-http01
- Eventing
  - in-memory channel (separable)
  - MT Channel based broker
  - Core eventing sources
- Tekton Pipelines
- CLI
  - Mink has a rudimentary CLI that doubles as a kn plug-in, which ties together some of the pieces.

**Sponsoring WG:** TBD.  

I'd be happy to put this under a new separate working group (think: sig-kind), or to sponsor it under the banner of Serving API (since it's largely an artifact of the controller architecture we've spearheaded).  I have a mild bias towards the former given the somewhat horizontal nature, but am curious what folks think?

cc @knative/api-core-wg-leads 
cc @knative/technical-oversight-committee 
cc @pmorie 

## Actions to fulfill

This area is used to track the [repo creation process](/mechanics/CREATING-A-SANDBOX-REPO.md).
The _requestor_ and _sponsoring WG lead_ should perform the steps listed below and cross out the checkmarks when done.
The TOC is involved only in the **TOC Gate** steps.

- [ ] Add this issue to the [TOC project board](https://github.com/orgs/knative/projects/9) for review.

_You may not be able to use the Projects quick menu on this page. In that case, go to the project board and use the **Add cards** interface._

- [ ] Send a PR adding entries for this repo in [`/peribolos/knative-sandbox.yaml`](https://github.com/knative/community/blob/master/peribolos/knative-sandbox.yaml). Please mind the alphabetical order when adding to a list.
  - [ ] Add the repository and a description.
  - [ ] Grant `Knative Admin` the `admin` privilege.
  - [ ] Grant the sponsoring WG the `write` privilege.

**TOC Gate**: _Once the TOC has approved the above, it will merge and Peribolos will create an empty repository._

- [ ] (golang) Send a PR to add aliases for `knative.dev/$REPONAME` import paths ([sample](https://github.com/knative/website/pull/187)).

- [ ] Have a lead from the sponsoring WG bootstrap the Git repository by pushing an
  appropriate "template" repository ([basic](https://github.com/knative-sandbox/wg-repository),
  [sample-controller](https://github.com/knative-sandbox/sample-controller),
  [sample-source](https://github.com/knative-sandbox/sample-source)) to the new repository as
  a git remote.  For example:

  ```shell
    git clone https://github.com/knative-sandbox/sample-controller.git
    cd sample-controller
    git remote add newrepo https://github.com/knative-sandbox/$REPONAME.git
    git push newrepo master
  ```

- [ ] Set up [test-infra](https://github.com/knative/test-infra) following the
  docs linked at the beginning.

- [ ] Create a sample PR to verify Prow (e.g. edit the boilerplate README)

**TOC Gate** : _Once Prow has been verified_

- [ ] Have a member of the TOC enable branch protections so that `tide` is a required presubmit check.

**Comment by evankanderson** [comment]
@pmorie , can you join the TOC meeting next week to square this with [your starter distribution proposal](https://docs.google.com/document/d/1A6Sji6Qd305VkJ1e0Yg1lECYA4ayKWJO0qesfrtZw3c/edit)?

**Comment by mattmoor** [comment]
> The goal is closer to the dynamic with KinD and minikube in Kubernetes (and to a lesser extent k3s under CNCF), to present an option, not the option

Just to reiterate this ☝️ 


It is notable that while I expect mink to be conformant (it passes all our tests), I don't want to limit it to the minimal set of things to achieve that conformance.  As discussed in the doc, if we had a minimally conformant starter-distro, I'd expect we might base mink on that (assuming it satisfies the mink design goals), but I would expect that starter-distro wouldn't include things like Tekton, which I feel is an important opinion mink holds.

Put differently, I would expect the scope of the starter-distro to be bound by trademark/conformance in ways that are express non-goals of mink, thus sandbox 😇

**Comment by grantr** [comment]
> Sponsoring WG: TBD.

IMO this is the central discussion point here. Independent of mink, this example brings up some interesting procedural questions for me.

- What do we do if there is no clear WG fit for a proposed repo?
  - Is it encouraged to propose a new WG to own the new repo?
  - Is the WG proposal and approval process the same as for any other WG?
  - Must the WG be approved before the repo can be created?
- What do we do if a WG wants to own a proposed repo, but that repo doesn't seem to fit with their charter?
  - What if another WG charter seems to be a better fit? (I think this came up in the buoy proposal #288)

**Comment by pmorie** [comment]
Notes from our discussion on Oct 29:

My personal opinion here: though mink is in proximity in some ways to the idea of the starter distribution, it’s also a lot ‘larger’ functionally and conceptually. I’m personally in favor of not blocking on a fully consensus’d, reified idea of a starter distribution, and in fact think it would be better to forge ahead with mink, because I think it will help us understand what the starter distribution should _be_.

A couple possibilities for where to home this project within the community:

- mink WG
- user experience WG

It seemed on the call people preferred the user experience WG route, of course if there are other ideas, please speak up!

**Comment by vaikas** [comment]
UX WG from my side. Sorry for terseness, on phone.

**Comment by mattmoor** [comment]
So I've been noodling on what a "User Experience" working group might encompass, and see some other things that might fit under its purview both in existing artifacts and in other external prototypes, but all having the aim of a better end-user experience.
1. The eventing "sugar" controller (auto-create brokers per-NS).
2. The serving "nscert" controller (auto-create wildcard kcerts per-NS).
3. @n3wscott's "auto-trigger" project

There are also several PoC's around auto-migration from Deployment to KService that might be interesting to consider under this umbrella.

**Comment by mattmoor** [comment]
Seems relevant: https://twitter.com/chrissjohnson00/status/1331088214958829568

**Comment by mattmoor** [comment]
cc @csantanapr @omerbensaadon (who have been talking about user-productivity / getting started)

**Comment by mattmoor** [comment]
Going to close this, and we can revisit under the new UX WG 🎉
